R/knitr-engine.R

```diff
@@ -492,19 +492,44 @@ eng_python_initialize_hooks <- function(options, envir) {
 
 }
 
+eng_python_matplotlib_backend <- function() {
+
+  # allow override, just in case
+  envvars <- c("RETICULATE_MPLBACKEND", "MPLBACKEND")
+  for (envvar in envvars) {
+    override <- Sys.getenv(envvar, unset = NA)
+    if (!is.na(override))
+      return(override)
+  }
+
+  # in RStudio Desktop, enforce a non-Qt matplotlib backend
+  # this is especially important with RStudio Desktop as attempting to use a Qt
+  # backend will cause issues due to mismatched Qt versions between RStudio and
+  # Anaconda environments, and will cause crashes when generating plots
+  if (is_rstudio_desktop())
+    return("agg")
+
+  # prefer using the agg backend in non-interactive environments
+  # (matplotlib might prefer using the tkAgg backend, but we've seen
+  # issues when trying to use in some environments, e.g. Windows)
+  #
+  # https://github.com/rstudio/rstudio/issues/13840
+  if (!interactive())
+    return("agg")
+
+  # otherwise, use whatever backend was already configured
+  ""
+}
+
 eng_python_initialize_matplotlib <- function(options, envir) {
 
   # early exit if we already initialized
   # (this onload hook is registered for multiple matplotlib submodules)
   if (identical(.globals$matplotlib_initialized, TRUE))
     return(TRUE)
 
-  # attempt to enforce a non-Qt matplotlib backend. this is especially important
-  # with RStudio Desktop as attempting to use a Qt backend will cause issues due
-  # to mismatched Qt versions between RStudio and Anaconda environments, and
-  # will cause crashes when attempting to generate plots
-  testthat <- Sys.getenv("TESTTHAT", unset = NA)
```

Comment:
Should we keep the check for if we're in a testthat context? I think we can run tests interactively too.

Reply:
Thanks for catching that -- I've brought that back.